### PR TITLE
Author guide updates

### PR DESCRIPTION
**Pull request by joe-lawrence**
A few minor updates to the author guide, including scrubbing kpatch.ko and the old stop_machine model from the doc:  link to upstream kernel API documentation where applicable.

### CODE REVIEW

**Comment by joe-lawrence** [comment]
note:  I just noticed some kpatch.ko specific info in the main README.md file, like the `KPATCH_FORCE_UNSAFE` macro... I can exorcise that stuff in another PR.

**Comment by joe-lawrence** [comment]
Added sections on constprop and sibling calls.

File: doc/patch-author-guide.md [hunk 1]
```diff
@@ -261,58 +224,76 @@ A shadow PID variable is allocated in `do_fork()`: it is associated with the
...
+@@ -1797,6 +1798,13 @@ long do_fork(unsigned long clone_flags,
+ 	if (!IS_ERR(p)) {
+ 		struct completion vfork;
+ 		struct pid *pid;
 +		int *newpid;
 +		static int ctr = 0;
 +
-+		newpid = kpatch_shadow_alloc(p, "newpid", sizeof(*newpid),
-+					     GFP_KERNEL);
++		newpid = klp_shadow_get_or_alloc(p, 0, sizeof(*newpid), GFP_KERNEL,
```
Comment (jpoimboe) [inline doc/patch-author-guide.md]: Maybe have `#define KPATCH_SHADOW_NEWPID 0` somewhere (or enum) and use that instead?  Perhaps along with a comment that the value needs to be globally unique.

File: doc/patch-author-guide.md [hunk 2]
```diff
@@ -324,9 +305,9 @@ structure can be used by patched code to handle both new (post-patch) and
 existing (pre-patch) instances.
 
 (This example is trimmed to highlight this use-case. Boilerplate code is also
-required to allocate/free a shadow variable called "reqs_active_v2" whenever a
-new `struct kioctx` is created/released. No values are ever assigned to the
-shadow variable.)
+required to allocate/free a shadow variable with enum ID SV_REQS_ACTIVE_V2
```
Comment (jpoimboe) [inline doc/patch-author-guide.md]: I had to do a double take at the SV_ prefix, maybe give it a KPATCH_SHADOW prefix instead?  Also it might be clearer to define it somewhere in the code.

File: doc/patch-author-guide.md [hunk 3]
```diff
@@ -386,7 +367,8 @@ the spinlock:
  		return NULL;
  
  	spin_lock_init(&sta->lock);
-+	ps_lock = kpatch_shadow_alloc(sta, "ps_lock", sizeof(*ps_lock), gfp);
++	ps_lock = klp_shadow_alloc(sta, PS_LOCK, sizeof(*ps_lock), gfp,
```
Comment (jpoimboe) [inline doc/patch-author-guide.md]: Similar comment here for PS_LOCK

File: doc/patch-author-guide.md [hunk 4]
```diff
@@ -193,66 +197,25 @@ static void kpatch_post_unpatch_tcp_send_challenge_ack(patch_object *obj)
...
+Don't forget to protect access to the data as needed. Spinlocks and mutexes /
+sleeping locks **may be used** (this is a change of behavior from when kpatch
+relied on the kpatch.ko support module and `stop_machine()` context.)
 
-2. If it can be assured that all patch targets will be loaded before the
-kpatch patch module, pre-patch callbacks may return -EBUSY if the lock/mutex
-is held to block the patching.
+Also note the pre-patch callback return code will be ignored by the kernel's
+module notifier, so it does not affect the target module or livepatch module
+status.
```
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Hmm, actually the original text is completely wrong.  The upstream documentation states:
>If a pre-patch callback returns failure, the kernel's module loader will:
>  - Refuse to load a livepatch, if the livepatch is loaded after targeted code.
>    or:
>  - Refuse to load a module, if the livepatch was already successfully loaded.

I'll update this accordingly.

File: doc/patch-author-guide.md [hunk 5]
```diff
@@ -261,58 +224,76 @@ A shadow PID variable is allocated in `do_fork()`: it is associated with the
 current `struct task_struct *p` value, given a string lookup key of "newpid",
 sized accordingly, and allocated as per `GFP_KERNEL` flag rules.
 
-`kpatch_shadow_alloc` returns a pointer to the shadow variable, so we can
+`klp_shadow_get_or_alloc()` returns a pointer to the shadow variable, so we can
```
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: FWIW, I had grabbed this from a local test/integration/fedora-30/shadow-newpid.patch, not realizing that it was never pushed up to the repository.  I think either call should work.  klp_shadow_get_or_alloc() is probably more straightforward.

File: doc/patch-author-guide.md (hunk 1 above)
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: ok

File: doc/patch-author-guide.md (hunk 2 above)
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: KPATCH_SHADOW_ prefix is kinda long.  Is something like ID_, KLP_SV, or KPS_ too cryptic?

File: doc/patch-author-guide.md [hunk 6]
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
...
+   f:   29 d6                   sub    %edx,%esi
+  11:   83 ee 14                sub    $0x14,%esi
+  ...
+```
+
+This could be a little confusing since one might have expected to see
+changes to all of `__tcp_mtu_to_mss()` callers (ie, it was inlined as
+requested).  In this case, a simple workaround is to provide a new,
+non-inlined copy of `__tcp_mtu_to_mss()` and then update all its callers
+to use it.
```
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Yeah I think that would work, too.

File: doc/patch-author-guide.md [hunk 7]
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
...
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.9
+```
+The current workaround for this issue to add `__attribute__((noclone))`
+to offending functions.
+
+Sibling calls
+-------------
+
+If function A() calls function B() and they share a common function
+signature, GCC may deem them "sibling calls" and apply a tail call
```
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Hmm, I suppose that is true, I haven't looked into the implementation of how they implement it.  As long as the stack expectations line up when B() is called, I can see how the signatures don't need to match exactly.  I can reword this to state that the compiler may optimize A() calling B() at the end of the function, skipping some calling conventions to save overhead.

File: doc/patch-author-guide.md [hunk 8]
```diff
@@ -597,3 +579,126 @@ For example, a `pr_warn_once()` can be replaced with:
...
+optimization in which the call to B() avoids the overhead of an ordinary
+function call.  This may be an issue for kpatches which only A() or B():
```
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Ah, yes "patch A() or B()".  I'll add a bit about ppc64le and "external" calls.

File: doc/patch-author-guide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jpoimboe) [reply doc/patch-author-guide.md]: How would klp_shadow_get_or_alloc() be more straightforward here?  In do_fork() there shouldn't be an existing shadow struct for it, right?

File: doc/patch-author-guide.md (hunk 2 above)
Comment (jpoimboe) [reply doc/patch-author-guide.md]: If it makes it more readable, why not go with the longer version?  The extra characters are free.  KLP_SV and KPS_ are cryptic, and ID_ doesn't make clear that it's a kpatch-specific id.

File: doc/patch-author-guide.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpoimboe) [reply doc/patch-author-guide.md]: The signatures don't need to match because for sibling calls, A() restores the stack to its callee's state, and then sets up B()'s args appropriately before jumping to B().

File: doc/patch-author-guide.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: D'oh.. copy/paste error.  I meant to say klp_shadow_alloc() for that reason.

[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Actually.... this bugged me as I found the same klp_shadow_get_or_alloc() usage in the rhel-7.x integration tests.  It traces back to the centos-7 tests in which @sm00th modified the test in 56be80720aee ("integration/centos-7: fix shadow-newpid.patch warnings").  From the commit message:
> When quickly loading/unloading this patch multiple times it is possible to hit "Duplicate shadow variable" warnings since the patch doesn't have any cleanup hooks on unload.
Switch to klp_shadow_get_or_alloc to ignore these.

I can provide the original version of the patch in the author guide using klp_shadow_alloc() and mention that it was adjusted for testing purposes or some other disclaimer.

[the remark this answers is not in the document]
Comment (jpoimboe) [reply doc/patch-author-guide.md]: Ah, that makes sense now.  I guess the correct fix would be to use klp_shadow_free_all() in a post-unload callback?  If so it might be worth adding that to the real integration patches as well for proper test coverage.

[the remark this answers is not in the document]
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: I would say that is the correct solution for a real kpatch, however I usually try hard not to mix features when writing test cases.  In this case, I think we could call klp_shadow_free_all() from a module_exit function to achieve the same result.

**Comment by joe-lawrence** [comment]
Another round of revisions.. I redid the shadow variable diff hunks to reflect the new API and @jpoimboe 's suggestions.  A few other nits found here and there (kpatch.ko references keep popping up the more I read the doc.)

File: doc/patch-author-guide.md [hunk 9]
```diff
@@ -680,3 +680,24 @@ if (static_key_true(&key1))
...
+```
+# eu-readelf -s orig/kernel/time/timekeeping.o  | grep timekeeping_update
+  137: 0000000000000000    552 FUNC    LOCAL  DEFAULT       48 timekeeping_update.constprop.9
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.8
+# eu-readelf -s patched/kernel/time/timekeeping.o  | grep timekeeping_update
+  137: 0000000000000000    552 FUNC    LOCAL  DEFAULT       48 timekeeping_update.constprop.10
+  138: 0000000000000000    568 FUNC    LOCAL  DEFAULT       50 timekeeping_update.constprop.9
+```
+The current workaround for this issue to add `__attribute__((noclone))`
+to offending functions.
```
Comment (jpoimboe) [inline doc/patch-author-guide.md]: I think this issue was fixed in #1043.  Should this commit be dropped?

File: doc/patch-author-guide.md [hunk 10]
```diff
@@ -701,3 +701,23 @@ correleating constprop subfunctions created by the optimization:
...
+the function call crosses a kernel module boundary: the sibling call
+optimization has changed expected calling conventions and (un)patched code may
+not be similarly modified.
+
+Commit [8b952bd77130](https://github.com/dynup/kpatch/commit/8b952bd77130)
+("create-diff-object/ppc64le: Don't allow sibling calls") contains an
+excellent example and description of this problem with annotated disassembly.
+
+Adding `__attribute__((optimize("-fno-optimize-sibling-calls")))` instructs
+GCC to turn off the optimization for the given function.
```
Comment (jpoimboe) [inline doc/patch-author-guide.md]: Reading this, it's not clear how the user should know they've hit this issue in the first place.  It should probably mention the error message:
```
			ERROR("Found an unsupported sibling call at %s()+0x%lx.  Add __attribute__((optimize(\"-fno-optimize-sibling-calls\"))) to %s() definition.",
```

File: doc/patch-author-guide.md (hunk 9 above)
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Yeah, I think it can go

File: doc/patch-author-guide.md (hunk 10 above)
Comment (joe-lawrence) [reply doc/patch-author-guide.md]: Good point, I'll add it to the beginning of the section.

**Comment by joe-lawrence** [comment]
Addressed @jpoimboe 's comments on the previous version and added a few more static key examples.  In between revisions, there have been a few new issues opened concerning the guide (callback best practices, etc.) that we can address with a follow up PR.  This one has been neglected long enough :)
